### PR TITLE
docs: add domain model of workspace management

### LINKED ISSUE #617: Feature: add workspace management domain models
### What would you like to be added?

Add workspace management domain models, including `WorkspaceConfig`, `Module`, `BackendConfig`, `RuntimeConfig`, etc.

### Why is this needed?

The first step to land workspace management.

### PR DESCRIPTION
**Pull request by healthjyk**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
 
add domain model of workspace management.

#### Which issue(s) this PR fixes:

Fixes #617 

#### Special notes for your reviewer:
The package registry is used to workspace config deserialization.

### CODE REVIEW

File: pkg/models/module.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+	ModuleName() string
+}
+
+// SecureWorkspaceModule is used to hold Module fields retrieved from workspace config which get
+// verified.
+type SecureWorkspaceModule interface {
+	Module
+
+	// ValidateWorkspaceModule validates the Module retrieved from workspace is correct or not.
+	ValidateWorkspaceModule() error
```
Comment (SparkYuan) [inline pkg/models/module.go:18]: Please rename `ValidateWorkspaceModule` to `Validate` to make it concise, considering it is already within the `SecureWorkspaceModule` definition.

File: pkg/models/module.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package models
+
+// Module is an interface to describe the domain model of a building block. A Module comes both
+// from an application developer-centric schema and workspace config, which can fully describes a workload,
+// an accessory, etc., such as database.Database, network.Port.
+// A Module can be built into a or a set of Resource, which will be consumed by Kusion engine.
+type Module interface {
+	// ModuleName returns the name to identify the module uniquely.
+	ModuleName() string
```
Comment (SparkYuan) [inline pkg/models/module.go:9]: Please rename `ModuleName` to `Name` to follow the Go convention

File: pkg/models/config.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
+package models
+
+// ModuleConfig is an interface to describe the config of a module, which is organized in the
```
Comment (SparkYuan) [inline pkg/models/config.go:3]: Whether ModuleConfig is an interface or a struct? The comment conflicts with the code

File: pkg/models/module.go (hunk 1 above)
Comment (healthjyk) [reply pkg/models/module.go:18]: Cause the module has two sources: 1, scheme provided by application developer; 2, workspace config provided by platform engineer. To do validation separately, there are functions `ValidateWorkspaceModule` and `ValidateSchemeModule`.

File: pkg/models/module.go (hunk 2 above)
Comment (healthjyk) [reply pkg/models/module.go:9]: The function `ModuleName` is used to define a module. If changed to `Name`, the capacity of definition will not exist.

File: pkg/models/config.go (hunk 3 above)
Comment (healthjyk) [reply pkg/models/config.go:3]: It's a struct. This is a miswriting.

File: pkg/workspace/config.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
+package workspace
+
+import (
+	"kusionstack.io/kusion/pkg/models"
+)
+
+// Config is used to describe the config of a workspace.
+type Config struct {
```
Comment (adohe) [inline pkg/workspace/config.go:8]: Why not just use `type Workspace struct` here? What's the difference between `Config` and `Workspace`?

File: pkg/registry/registry.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+// config structure, so that the workspace config can be correctly deserialized.
+package registry
+
+import (
+	"kusionstack.io/kusion/pkg/models"
+	"kusionstack.io/kusion/pkg/models/appconfiguration/accessories/database"
+)
+
+var (
+	registeredModules  []models.Module
```
Comment (adohe) [inline pkg/registry/registry.go:11]: The module code might be put out side of Kusion repo, so how to register modules here?

File: pkg/workspace/operator.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package workspace
```
Comment (adohe) [inline pkg/workspace/operator.go:1]: Personally I would suggest you only include API types definition in this PR, operator code can be a separate PR.

Comment (healthjyk) [reply pkg/workspace/operator.go:1]: That's cool and I'll change it.

File: pkg/registry/registry.go (hunk 5 above)
Comment (healthjyk) [reply pkg/registry/registry.go:11]: The registry pattern is based on the assumption that Kusion can get all the module structure. For the module code  put outside kusion repo, the thing goes the same, kusion has to get the module structure, whatever the way.
Of course, there is more work, and the exact work is unknown. It depends on the new generator scheme.
Hence, it seems that forgetting about the registry, and make it simple is more suitable for now.

**Comment by healthjyk** [comment]
This pr get closed, upgrade is https://github.com/KusionStack/kusion/pull/639
